daphne/src/taskprov.rs

```diff
@@ -0,0 +1,171 @@
+    },
+    vdaf::VdafVerifyKey,
+    DapAbort, DapError, DapQueryConfig, DapTaskConfig, DapVersion, Prio3Config, VdafConfig,
+};
+use prio::codec::Decode;
+use ring::{
+    hkdf::{Prk, Salt, HKDF_SHA256},
+    hmac::HMAC_SHA256,
+    hmac::{sign, Key},
+};
+use std::str;
+use url::Url;
+
+/// SHA-256 of "dap-taskprov-00"
+#[allow(dead_code)]
+pub(crate) const TASK_PROV_SALT: [u8; 32] = [
+    0x4d, 0x63, 0x1a, 0xeb, 0xa8, 0xdf, 0xe0, 0x1b, 0x34, 0x4c, 0x29, 0x2d, 0x17, 0xba, 0x34, 0x9a,
+    0x78, 0x97, 0xbf, 0x64, 0x88, 0x00, 0x55, 0x1c, 0x0d, 0x75, 0x32, 0xab, 0x61, 0x4b, 0xe2, 0x21,
+];
+
+/// Compute the task id of a serialized task config.
+pub fn compute_task_id(serialized: &[u8]) -> Id {
+    let key = Key::new(HMAC_SHA256, &TASK_PROV_SALT);
+    let digest = sign(&key, serialized);
+    let mut b: [u8; 32] = [0; 32];
+    let d = digest.as_ref();
+    b[..32].copy_from_slice(&d[..32]);
+    Id(b)
+}
+
+/// Compute the VDAF pseudorandom key for the pre-shared secret verify_key_init.
+pub(crate) fn compute_vdaf_verify_prk(verify_key_init: &[u8]) -> Prk {
+    // The documentation says computing the Salt is expensive, and we use the same PRK all the
+    // time, so we compute it once.
+    Salt::new(HKDF_SHA256, &TASK_PROV_SALT).extract(verify_key_init)
+}
+
+/// Compute the VDAF verify key for task_id and the specified VDAF type using the
+/// specified pseudorandom key.
+pub(crate) fn compute_vdaf_verify_key_from_prk(
+    prk: &Prk,
+    task_id: &Id,
+    vdaf_type: VdafType,
+) -> VdafVerifyKey {
+    let info = [task_id.as_ref()];
+    match &vdaf_type {
+        VdafType::Prio3Aes128Count | VdafType::Prio3Aes128Sum | VdafType::Prio3Aes128Histogram => {
+            // This expand(), and the associated fill() below can only fail if the length is wrong,
+            // and it won't be, so we unwrap().
+            let okm = prk.expand(&info, vdaf_type.clone()).unwrap();
+            let mut bytes = [0u8; 16];
+            okm.fill(&mut bytes[..]).unwrap();
+            VdafVerifyKey::Prio3(bytes)
+        }
+        _ => panic!("Unknown VDAF type"),
+    }
+}
+
+/// Compute the VDAF verify key for task_id and the specified VDAF type using the
+/// pre-shared secret verify_key_init.
+///
+/// This is a convenience function to call compute_vdaf_verify_prk() and
+/// compute_vdaf_verify_key_from_prk().  Callers reusing the same PRK frequently
+/// should consider computing the prk once and then calling compute_vdaf_verify_key_from_prk()
+/// directly.
+#[allow(dead_code)]
+pub(crate) fn compute_vdaf_verify_key(
+    verify_key_init: &[u8],
+    task_id: &Id,
+    vdaf_type: VdafType,
+) -> VdafVerifyKey {
+    compute_vdaf_verify_key_from_prk(
+        &compute_vdaf_verify_prk(verify_key_init),
+        task_id,
+        vdaf_type,
+    )
+}
+
+pub fn bad_request(detail: &str) -> DapError {
+    DapError::Abort(DapAbort::BadRequest(detail.to_string()))
+}
+
+/// Check for a taskprov extension in the report, and return it if found.
```

Comment:
Parsing issues should generally be handled as abort "unrecognizedMessage". In particular, I think it would be better to parse the extension in the implementation of `Decode` for `Extension` (messages/mod.rs)`.

However you have hit on another point here that Tim G. noticed today: https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/375. We'll  have to decide how to address this in 03, but for now I think it makes perfect sense to just treat duplicate extensions as decoding failures. In particular, we should lift this de-duplication step to the implementation of `Decode` on `ReportMetadata`.

Reply:
good idea, will do